library/x509_create.c

```diff
@@ -86,62 +92,141 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
+    if( c == end || *c != ',' )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+    }
+
+    return( ret );
+}
+
+static int dn_store_named_data( mbedtls_asn1_named_data **head, const char *oid,
+                                char data[MBEDTLS_X509_MAX_DN_NAME_SIZE], char *d,
+                                const char **s, x509_dn_parse_state *dn_cur_state,
+                                const char **cur, const char *end )
+{
+    const char *c = *cur;
+
+    if( mbedtls_asn1_store_named_data( head, oid, strlen( oid ),
+                                       (unsigned char *) data,
+                                       d - data ) == NULL )
+    {
+        return( MBEDTLS_ERR_X509_ALLOC_FAILED );
+    }
+
+    /*
+     * Skip white spaces
+     */
+    while( c < end && *(c + 1) == ' ' )
+        c++;
+
+    *s = c + 1;
+    *dn_cur_state = dn_in_tag_no_quot;
+
+    return ( 0 );
+}
+
 int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *name )
 {
     int ret = 0;
     const char *s = name, *c = s;
     const char *end = s + strlen( s );
     const char *oid = NULL;
-    int in_tag = 1;
     char data[MBEDTLS_X509_MAX_DN_NAME_SIZE];
     char *d = data;
+    x509_dn_parse_state dn_cur_state = dn_in_tag_no_quot;
 
     /* Clear existing chain if present */
     mbedtls_asn1_free_named_data_list( head );
 
     while( c <= end )
     {
-        if( in_tag && *c == '=' )
+        switch( dn_cur_state )
         {
-            if( ( oid = x509_at_oid_from_name( s, c - s ) ) == NULL )
+            case dn_in_tag_no_quot:
             {
-                ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
-                goto exit;
+                if( *c == '=' )
+                {
+                    if( ( oid = x509_at_oid_from_name( s, c - s ) ) == NULL )
+                    {
+                        ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
+                        goto exit;
+                    }
+                    s = c + 1;
+                    d = data;
+                    if( *s == '\"' )
+                    {
+                        dn_cur_state = dn_no_tag_in_quot;
+                        s++;
+                        c += 2;
+                    }
+                    else
+                    {
+                        dn_cur_state = dn_no_tag_no_quot;
+                    }
+                }
             }
+            break;
 
-            s = c + 1;
-            in_tag = 0;
-            d = data;
-        }
-
-        if( !in_tag && *c == '\\' && c != end )
-        {
-            c++;
-
-            /* Check for valid escaped characters */
-            if( c == end || *c != ',' )
+            case dn_no_tag_in_quot:
             {
-                ret = MBEDTLS_ERR_X509_INVALID_NAME;
-                goto exit;
+                if( *c == '\"' && c != end )
+                {
+                    c++;
+                    s = c + 1;
+                    if( *c != ',' )
```

Comment:
I'll skip whitespaces

Reply:
Reading the RFC again, only one optional space is allowed. I will check only one space

Reply:
@RonEld I disagree, the RFC contains the following grammar snippet:
```
<spaced-separator> ::= <optional-space>
                   <separator>
                   <optional-space>

   <separator> ::=  "," | ";"

   <optional-space> ::= ( <CR> ) *( " " )
```

Reply:
See [RFC 822, Sect. 2.4](https://tools.ietf.org/html/rfc822#section-2).

Reply:
>Doesn't the `*` mean that `" "` may occur an arbitrary number of times?

Perhaps, when I read the RFC yesterday I understood it could happen only once. I'll skip all whitespaces